interchain-token-service/src/contract/execute.rs

```diff
@@ -0,0 +1,330 @@
+use axelar_wasm_std::IntoContractError;
+use cosmwasm_std::{DepsMut, HexBinary, QuerierWrapper, Response, Storage};
+use error_stack::{bail, ensure, report, Result, ResultExt};
+use router_api::{Address, ChainName, ChainNameRaw, CrossChainId};
+
+use crate::events::ItsContractEvent;
+use crate::primitives::ItsHubMessage;
+use crate::state::{self, load_config, load_its_address};
+
+#[derive(thiserror::Error, Debug, IntoContractError)]
+pub enum Error {
+    #[error("unknown chain {0}")]
+    UnknownChain(ChainName),
+    #[error("unable to load the contract config")]
+    ConfigAccess,
+    #[error("unknown its address {0}")]
+    UnknownItsAddress(Address),
+    #[error("failed to decode payload")]
+    InvalidPayload,
+}
+
+/// Executes an incoming ITS message.
+///
+/// This function handles the execution of ITS (Interchain Token Service) messages received from
+/// its sources. It verifies the source address, decodes the message, applies balance tracking,
+/// and forwards the message to the destination chain.
+pub fn execute_message(
+    deps: DepsMut,
+    cc_id: CrossChainId,
+    source_address: Address,
+    payload: HexBinary,
+) -> Result<Response, Error> {
+    ensure_its_source_address(deps.storage, &cc_id.source_chain, &source_address)?;
+
+    match ItsHubMessage::abi_decode(&payload).change_context(Error::InvalidPayload)? {
+        ItsHubMessage::SendToHub {
+            destination_chain,
+            message,
+        } => {
+            let destination_chain = normalize(&destination_chain);
+            let destination_address = load_its_address(deps.storage, &destination_chain)
+                .change_context_lazy(|| Error::UnknownChain(destination_chain.clone()))?;
+
+            let destination_payload = ItsHubMessage::ReceiveFromHub {
```

Comment:
Why is the source address not encoded in the payload?

Reply:
The `source_address` is of the ITS edge contract on source chain which is not relevant for the destination. The real `source_address` corresponding to the caller of ITS is inside the [its message](https://github.com/axelarnetwork/axelar-amplifier/blob/e38843b9bb62c79c0038ef96d24df559f5960157/interchain-token-service/src/primitives.rs#L41). Since ITS is an app on top of GMP, we can't get around this wrapping issue.